pkg/declarative/v2/reconciler.go

```diff
@@ -0,0 +1,405 @@
+		meta.SetStatusCondition(&status.Conditions, condition)
+	}
+	status.Synced = make([]Resource, 0)
+	obj.SetStatus(status)
+}
+
+func (r *Reconciler) initializeClient(
+	ctx context.Context, obj Object, spec *Spec,
+) (Client, error) {
+	var err error
+	var clnt Client
+
+	clientsCacheKey := cacheKeyFromObject(ctx, obj)
+
+	if clnt = r.GetClients(clientsCacheKey); clnt == nil {
+		cluster := &types.ClusterInfo{
+			Config: r.Config,
+			Client: r.Client,
+		}
+		if r.TargetClient != nil {
+			cluster.Client = r.TargetClient
+		}
+		clnt, err = manifestClient.NewSingletonClients(cluster, log.FromContext(ctx))
+		if err != nil {
+			return nil, err
+		}
+		clnt.Install().Atomic = false
+		clnt.Install().Replace = true
+		clnt.Install().DryRun = true
+		clnt.Install().IncludeCRDs = false
+		clnt.Install().CreateNamespace = r.CreateNamespace
+		clnt.Install().UseReleaseName = false
+		clnt.Install().IsUpgrade = true
+		clnt.Install().DisableHooks = true
+		clnt.Install().DisableOpenAPIValidation = true
+		if clnt.Install().Version == "" && clnt.Install().Devel {
+			clnt.Install().Version = ">0.0.0-0"
+		}
+		clnt.Install().ReleaseName = spec.ManifestName
+		r.SetClients(clientsCacheKey, clnt)
+	}
+
+	clnt.Install().Namespace = r.Namespace
+	clnt.KubeClient().Namespace = r.Namespace
+
+	if r.Namespace != metav1.NamespaceNone && r.Namespace != metav1.NamespaceDefault &&
+		clnt.Install().CreateNamespace {
+		err := clnt.Patch(
+			ctx, &v1.Namespace{
+				TypeMeta:   metav1.TypeMeta{APIVersion: "v1", Kind: "Namespace"},
+				ObjectMeta: metav1.ObjectMeta{Name: r.Namespace},
+			}, client.Apply, client.ForceOwnership, r.FieldOwner,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return clnt, nil
+}
+
+func cacheKeyFromObject(ctx context.Context, resource client.Object) client.ObjectKey {
+	logger := log.FromContext(ctx)
+
+	if resource == nil {
+		return client.ObjectKey{}
+	}
+
+	label, err := util.GetResourceLabel(resource, manifestLabels.CacheKey)
+	objectKey := client.ObjectKeyFromObject(resource)
+	var labelErr *types.LabelNotFoundError
+	if errors.As(err, &labelErr) {
+		logger.V(util.DebugLogLevel).Info(
+			manifestLabels.CacheKey+" missing on resource, it will be cached "+
+				"based on resource name and namespace.",
+			"resource", objectKey,
+		)
+		return objectKey
+	}
+
+	logger.V(util.DebugLogLevel).Info(
+		"resource will be cached based on "+manifestLabels.CacheKey,
+		"resource", objectKey,
+		"label", manifestLabels.CacheKey,
+		"labelValue", label,
+	)
+
+	return client.ObjectKey{Name: label, Namespace: resource.GetNamespace()}
+}
+
+func (r *Reconciler) ssaStatus(ctx context.Context, obj client.Object) (ctrl.Result, error) {
+	obj.SetResourceVersion("")
+	return ctrl.Result{Requeue: true}, r.Status().Patch(ctx, obj, client.Apply, client.ForceOwnership, r.FieldOwner)
+}
+
+func (r *Reconciler) ssa(ctx context.Context, obj client.Object) (ctrl.Result, error) {
+	obj.SetResourceVersion("")
```

Comment:
do we need this? 
AFAIK only managed fields should be nil for an object with SSA

Reply:
ResourceVersion is managed and will result in an optimistic lock.